heartbeat/monitors/util.go

```diff
@@ -370,10 +372,13 @@ func resolveErr(host string, err error) (common.MapStr, []TaskRunner, error) {
 func WithFields(fields common.MapStr, r TaskRunner) TaskRunner {
 	return MakeCont(func() (common.MapStr, []TaskRunner, error) {
 		event, cont, err := r.Run()
-		if event == nil {
+		if event != nil {
+			event = event.Clone()
+			event.DeepUpdate(fields)
+		} else if err != nil {
 			event = common.MapStr{}
+			event.DeepUpdate(fields)
```

Comment:
Reading this with the `else if` I asked myself what the `else` case in the end which is not covered here actually means and if it can happen. So event is nil and err is nil?

Reply:
A runner can return an event, but must not. It's fully optional. We have intermediate runners (resolve IP address), that do not return an event, but create a set of continuation runners. So `event` can be nil. 

If a runner returns an error, no followup runners are created. => We need to create an error event. On error a runner can optionally return an event. The WithFields runner enforces the creation of an event on error, such that the already known fields (like IP resolve results) an be merged into the final event

The WithFields wrapper adds shared fields only if an event will be returned.

To be more verbose, the conditions is:

```
if err != nil {
  if event == nil {
    event = common.MapStr{}
  }
  event.DeepUpdate(fields)
} else { // no error
  if event != nil {
    event = event.Clone() // we need to copy the event to ensure no error are returned.
    event.DeepUpdate(fields)
  }
}
```

Due to the `else if`, the second else branch is equivalent to `event == nil && err != nil`.

The case `event == nil && err == nil` is perfectly valid -> we just don't report an event.